frontend/src/js/previous-queries/list/useDeleteProjectItemFolder.tsx

```diff
@@ -0,0 +1,86 @@
+import { StateT } from "app-types";
+import { useTranslation } from "react-i18next";
+import { useSelector, useDispatch } from "react-redux";
+
+import { useDatasetId } from "../../dataset/selectors";
+import { setMessage } from "../../snack-message/actions";
+
+import {
+  removeFolder,
+  useLoadFormConfigs,
+  useLoadQueries,
+  useUpdateFormConfig,
+  useUpdateQuery,
+} from "./actions";
+import type { FormConfigT, PreviousQueryT } from "./reducer";
+
+export const useDeleteProjectItemFolder = (
+  folder: string,
+  onSuccess?: () => void,
+) => {
+  const { t } = useTranslation();
+  const datasetId = useDatasetId();
+  const dispatch = useDispatch();
+  const { updateQuery } = useUpdateQuery();
+  const { updateFormConfig } = useUpdateFormConfig();
+  const { loadQueries } = useLoadQueries();
+  const { loadFormConfigs } = useLoadFormConfigs();
+
+  const queries = useSelector<StateT, PreviousQueryT[]>(
+    (state) => state.previousQueries.queries,
+  );
+  const formConfigs = useSelector<StateT, FormConfigT[]>(
+    (state) => state.previousQueries.formConfigs,
+  );
+
+  const localFolders = useSelector<StateT, string[]>(
+    (state) => state.previousQueries.localFolders,
+  );
+
+  return async () => {
+    if (!datasetId) return;
+
+    if (localFolders.includes(folder)) {
+      dispatch(removeFolder({ folderName: folder }));
+      if (onSuccess) {
+        onSuccess();
+      }
+      return;
+    }
+
+    try {
+      await Promise.all([
+        ...queries
+          .filter((query) => query.tags.includes(folder))
+          .map((query) => {
+            const nextTags = query.tags.filter((tag) => tag !== folder);
+
+            return updateQuery(
+              query.id,
+              { tags: nextTags },
+              t("previousQuery.retagError"),
+            );
+          }),
+        ...formConfigs
+          .filter((config) => config.tags.includes(folder))
+          .map((config) => {
+            const nextTags = config.tags.filter((tag) => tag !== folder);
+
+            return updateFormConfig(
+              config.id,
+              { tags: nextTags },
+              t("formConfig.retagError"),
+            );
+          }),
+      ]);
+
+      await Promise.all([loadQueries(datasetId), loadFormConfigs(datasetId)]);
+
+      if (onSuccess) {
+        onSuccess();
+      }
+    } catch (e) {
+      dispatch(setMessage({ message: t("previousQuery.retagError") }));
+    }
```

Comment:
Would something like this be possible? I'm a little confused that you are not using the return values of your methods. But I guess you exchange the data using redux?
```suggestion
      await Promise.all([
        ...queries
          .filter((query) => query.tags.includes(folder))
          .map((query) => {
            const nextTags = query.tags.filter((tag) => tag !== folder);

            return updateQuery(
              query.id,
              { tags: nextTags },
              t("previousQuery.retagError"),
            );
          }),
        ...formConfigs
          .filter((config) => config.tags.includes(folder))
          .map((config) => {
            const nextTags = config.tags.filter((tag) => tag !== folder);

            return updateFormConfig(
              config.id,
              { tags: nextTags },
              t("formConfig.retagError"),
            );
          }),
      ])
      .then( _ => Promise.all([loadQueries(datasetId), loadFormConfigs(datasetId)]))
      .then( _ => onSuccess())
      .catch( err => dispatch(setMessage({ message: t("previousQuery.retagError") })));
```

Reply:
then / catch was used back in the days before async / await control flow existed. I'm not sure I understand your confusion :)